src/parsers/results.ts

```diff
@@ -0,0 +1,228 @@
+    })
+    observations.push({
+      date: new Date(interval),
+      value: intervals[interval].observations,
+    })
+  }
+
+  findings.sort((a, b) => {
+    return a.date.valueOf() - b.date.valueOf()
+  })
+  observations.sort((a, b) => {
+    return a.date.valueOf() - b.date.valueOf()
+  })
+
+  return {
+    datasets: [
+      {
+        label: 'Findings',
+        gradient: {
+          backgroundColor: {
+            axis: 'y',
+            colors: {
+              100: 'rgba(253,92,110,0.4)',
+              70: 'rgba(253,92,110, .3)',
+              30: 'rgba(253,92,110, .1)',
+              0: 'rgba(253,92,110, .0)',
+            },
+          },
+        },
+        borderColor: 'rgba(253,92,110, 0.7)',
+        data: findings,
+        parsing: {
+          xAxisKey: 'date',
+          yAxisKey: 'value',
+        },
+      },
+      {
+        label: 'Observations',
+        gradient: {
+          backgroundColor: {
+            axis: 'y',
+            colors: {
+              100: 'rgba(20,184,166, .4)',
+              70: 'rgba(20,184,166, .3)',
+              30: 'rgba(20,184,166, .1)',
+              0: 'rgba(20,184,166, .0)',
+            },
+          },
+        },
+        borderColor: 'rgba(20,184,166, 0.7)',
+        data: observations,
+        parsing: {
+          xAxisKey: 'date',
+          yAxisKey: 'value',
+        },
+      },
+    ],
+  }
+}
+
+export function calculateAgentUptimeData(
+  results: ComplianceBySearchResult[],
+): ChartData<'line', DateDataPoint[]> {
+  const intervals: {
+    [interval: string]: {
+      uptime: number
+    }
+  } = {}
+
+  for (const result of results) {
+    for (const record of result.records) {
+      if (intervals.hasOwnProperty(record.interval)) {
+        intervals[record.interval].uptime =
+          intervals[record.interval].uptime + (record.hasRecords ? 1 : 0)
+      } else {
+        intervals[record.interval] = {
+          uptime: record.hasRecords ? 1 : 0,
+        }
+      }
+    }
+  }
+
+  const dataList = []
+  for (const interval in intervals) {
+    dataList.push({
+      date: new Date(interval),
+      value: intervals[interval].uptime,
+    })
+  }
+
+  dataList.sort((a, b) => {
+    return a.date.valueOf() - b.date.valueOf()
+  })
+
+  return {
+    datasets: [
+      {
+        label: 'HeartBeats',
+        gradient: {
+          backgroundColor: {
+            axis: 'y',
+            colors: {
+              100: 'rgba(20,184,166, .4)',
+              70: 'rgba(20,184,166, .3)',
+              30: 'rgba(20,184,166, .1)',
+              0: 'rgba(20,184,166, .0)',
+            },
+          },
+        },
+        borderColor: 'rgba(20,184,166, 0.7)',
+        data: dataList,
+        parsing: {
+          xAxisKey: 'date',
+          yAxisKey: 'value',
+        },
+      },
+    ],
+  } as ChartData<'line', DateDataPoint[]>
+}
+
+export function calculateComplianceRatioData(results: Result[]): ChartData<'line', number[]> {
```

Comment:
Don't mutate when you can just declare instead.

```suggestion
export function calculateComplianceRatioData(results: Result[]): ChartData<'line', number[]> {
  const findings = results.reduce((total, result) => total + result.findings.length, 0);
  const observations = results.reduce((total, result) => total + result.observations.length, 0);

  return {
    labels: ['Findings', 'Observations'],
    datasets: [
      {
        backgroundColor: ['rgb(253,92,110)', 'rgb(20,184,166)'],
        data: [findings, observations],
      },
    ],
  } as ChartData<'line', number[]>;
}
```

Reply:
Done